File: backend/src/main/java/info/gamewise/lor/videos/deckcodeextractor/DescriptionWithRegisteredDeckExtractor.java
package info.gamewise.lor.videos.deckcodeextractor;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

import java.io.Serializable;
import java.util.Optional;

import static java.util.Collections.singletonList;
import static java.util.Optional.ofNullable;
import static org.springframework.http.HttpMethod.GET;
import static org.springframework.http.MediaType.APPLICATION_JSON;

public class DescriptionWithRegisteredDeckExtractor extends AbstractDeckCodeExtractor implements DeckCodeExtractor {

    private static final String URL_MOBA = "https://lor.mobalytics.gg/";
    private static final String URL_DECKS = URL_MOBA + "decks";
    private static final String URL_DECK_BY_ID = URL_MOBA + "api/v2/decks/library/by-id/";

    public static void main(String[] args) {
        DeckCodeExtractor extractor = new DescriptionWithRegisteredDeckExtractor();
        Optional<String> code = extractor.extract("https://lor.mobalytics.gg/decks/c18gl41509nhf1e6st80");
        code.ifPresentOrElse(System.out::println, () -> System.out.println("no deck code!"));
    }

    @Override
    public Optional<String> extract(String videoDescription) {
        return descriptionAsStream(videoDescription)
                .filter(this::filterMobalyticsLink)
                .findFirst()
                .map(this::deckCodeFromRegisteredDeck);
    }

    private String deckCodeFromRegisteredDeck(String url) {
        String deckId = url.substring(url.lastIndexOf("/") + 1);
        ResponseEntity<DeckResponse> response = requestMobalytics(URL_DECK_BY_ID + deckId);

        return ofNullable(response.getBody())
                .map(DeckResponse::getExportUID)
                .orElse(null);
    }

    private ResponseEntity<DeckResponse> requestMobalytics(String url) {
        RestTemplate restTemplate = new RestTemplate();
        return restTemplate.exchange(url, GET, httpEntity(), DeckResponse.class);
    }

    private HttpEntity<?> httpEntity() {
        HttpHeaders headers = new HttpHeaders();
        headers.setAccept(singletonList(APPLICATION_JSON));
        headers.setContentType(APPLICATION_JSON);
        return new HttpEntity<>(headers);
    }

    private boolean filterMobalyticsLink(String word) {
        return word.contains(URL_DECKS) && !word.contains("code");
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class DeckResponse implements Serializable {
        private String exportUID;

        public String getExportUID() {
            return exportUID;
        }

        public void setExportUID(String exportUID) {
            this.exportUID = exportUID;
        }
    }
}

File: backend/src/main/kotlin/info/gamewise/lor/videos/deckcodeextractor/DescriptionWithRegisteredDeckExtractor.kt
package info.gamewise.lor.videos.deckcodeextractor

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import org.slf4j.LoggerFactory
import org.springframework.http.*
import org.springframework.stereotype.Component
import org.springframework.web.client.RestTemplate
import java.io.Serializable
import java.util.Optional.ofNullable

private const val URL_MOBA = "https://lor.mobalytics.gg/"
private const val URL_DECKS = "${URL_MOBA}decks"
private const val URL_DECK_BY_ID = "${URL_MOBA}api/v2/decks/library/by-id/"

@Component
class DescriptionWithRegisteredDeckExtractor : AbstractDeckCodeExtractor(), DeckCodeExtractor {

    private val log = LoggerFactory.getLogger(DescriptionWithRegisteredDeckExtractor::class.java)

    override fun extract(videoDescription: String): String? {
        val mobaUrl = descriptionAsStream(videoDescription)
            .firstOrNull { filterMobalyticsLink(it) }
        return deckCodeFromRegisteredDeck(mobaUrl)
    }

    private fun filterMobalyticsLink(word: String): Boolean {
        return word.contains(URL_DECKS) && !word.contains("code")
    }

    private fun deckCodeFromRegisteredDeck(url: String?): String? {
        if (url == null) {
            return null
        }

        val deckId = url.substring(url.lastIndexOf("/") + 1)
        val response = requestMobalytics(URL_DECK_BY_ID + deckId)

        if (response.statusCode != HttpStatus.OK) {
            log.error("Deck not found for the URL={}", url)
            return null
        }

        return ofNullable(response.body)
            .map { it.exportUID }
            .orElse(null)
    }

    private fun requestMobalytics(url: String): ResponseEntity<DeckResponse> {
        return try {
            val restTemplate = RestTemplate()
            restTemplate.exchange(url, HttpMethod.GET, httpEntity(), DeckResponse::class.java)
        } catch (e: Exception) {
            ResponseEntity.notFound().build()
        }
    }

    private fun httpEntity(): HttpEntity<*> {
        val headers = HttpHeaders()
        headers.accept = listOf(MediaType.APPLICATION_JSON)
        headers.contentType = MediaType.APPLICATION_JSON
        return HttpEntity<Any>(headers)
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private class DeckResponse(var exportUID: String? = null) : Serializable

}